larpandoracontent/LArThreeDReco/LArTwoViewMatching/TwoViewThreeDKinkTool.cc

```diff
@@ -0,0 +1,480 @@
+/**
+ *  @file   larpandoracontent/LArThreeDReco/LArTwoViewMatching/TwoViewThreeDKinkTool.cc
+ *
+ *  @brief  Implementation of the three d kink base tool class.
+ *
+ *  $Log: $
+ */
+
+#include "Pandora/AlgorithmHeaders.h"
+
+#include "larpandoracontent/LArHelpers/LArClusterHelper.h"
+#include "larpandoracontent/LArHelpers/LArGeometryHelper.h"
+#include "larpandoracontent/LArObjects/LArPointingCluster.h"
+#include "larpandoracontent/LArHelpers/LArPointingClusterHelper.h"
+
+#include "larpandoracontent/LArThreeDReco/LArTwoViewMatching/TwoViewThreeDKinkTool.h"
+
+using namespace pandora;
+
+namespace lar_content
+{
+
+TwoViewThreeDKinkTool::TwoViewThreeDKinkTool() :
+    m_minXOverlapFraction(0.1f),
+    m_minMatchingScore(0.95f),
+    m_minLocallyMatchedFraction(0.3f),
+    m_minLongitudinalImpactParameter(-1.f),
+    m_nLayersForKinkSearch(10),
+    m_additionalXStepForKinkSearch(0.01f),
+    m_splitMode(false),
+    m_maxTransverseImpactParameter(5.f),
+    m_minImpactParameterCosTheta(0.5f),
+    m_cosThetaCutForKinkSearch(0.75f)
+{
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+TwoViewThreeDKinkTool::~TwoViewThreeDKinkTool()
+{
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+bool TwoViewThreeDKinkTool::Run(TwoViewTransverseTracksAlgorithm *const pAlgorithm, MatrixType &overlapMatrix)
+{
+    if (PandoraContentApi::GetSettings(*pAlgorithm)->ShouldDisplayAlgorithmInfo())
+        std::cout << "----> Running Algorithm Tool: " << this->GetInstanceName() << ", " << this->GetType() << std::endl;
+
+    ModificationList modificationList;
+    this->GetModifications(pAlgorithm, overlapMatrix, modificationList);
+    const bool changesMade(this->ApplyChanges(pAlgorithm, modificationList));
+
+    return changesMade;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+bool TwoViewThreeDKinkTool::PassesElementCuts(MatrixType::ElementList::const_iterator eIter, const ClusterSet &usedClusters) const
+{
+    if (usedClusters.count(eIter->GetCluster1()) || usedClusters.count(eIter->GetCluster2()))
+        return false;
+
+    if (eIter->GetOverlapResult().GetTwoViewXOverlap().GetXOverlapFraction0() - m_minXOverlapFraction < -1.f * std::numeric_limits<float>::epsilon())
```

Comment:
Hopefully the compiler handles this anyway, but `-std::numeric_limits<float>::epsilon()` will ensure the multiplication isn't actually performed at runtime (here and elsewhere)